src/vanilla/store.ts

```diff
@@ -355,35 +355,24 @@ export const createStore = () => {
     return nextAtomState
   }
 
-  const readAtomState = <Value>(atom: Atom<Value>): AtomState<Value> => {
+  const readAtomState = <Value>(
+    atom: Atom<Value>,
+    force?: boolean
+  ): AtomState<Value> => {
     // See if we can skip recomputing this atom.
     const atomState = getAtomState(atom)
-    if (atomState) {
-      // Ensure that each atom we depend on is up to date.
-      // Recursive calls to `readAtomState(a)` will recompute `a` if
-      // it's out of date thus increment its revision number if it changes.
-      atomState.d.forEach((_, a) => {
-        if (a !== atom && !mountedMap.has(a)) {
-          // Dependency is new or unmounted.
-          // Recomputing doesn't touch unmounted atoms, so we need to recurse
-          // into this dependency in case it needs to update.
-          readAtomState(a)
-        }
-      })
-      // If a dependency changed since this atom was last computed,
-      // then we're out of date and need to recompute.
+    if (!force && atomState) {
+      // If the atom is mounted, we can use the cache.
+      // because it should have been updated by dependencies.
+      if (mountedMap.has(atom)) {
+        return atomState
+      }
+      // Otherwise, check if the dependencies have changed.
+      // If all dependencies havne't changed, we can use the cache.
       if (
-        Array.from(atomState.d).every(([a, s]) => {
-          const aState = getAtomState(a)
-          return (
-            a === atom ||
-            aState === s ||
-            // TODO This is a hack, we should find a better solution.
-            (aState &&
-              !hasPromiseAtomValue(aState) &&
-              isEqualAtomValue(aState, s))
-          )
-        })
+        Array.from(atomState.d).every(
+          ([a, s]) => a === atom || readAtomState(a) === s
+        )
       ) {
         return atomState
```

Comment:
I hope to remove this for #1901, but this breaks various behavior including async atoms on mount. Hope to revisit it in the far future.